### DISCUSSION TITLE
Matplotlib Backend fails using quatro preview and render

### DISCUSSION THREAD

**Discussion by enryH** [quarto-dev/quarto-cli#3145] [Q&A]
I try to create my first quatro documents. I can execute the `qmd` notebook in VSCode ("Run" cell), but I cannot manage to make a preview. Somehow the backend is switched and I cannot figure out why. I tried the current release candidate (`1.2.256`) and the stable version ( `1.1.x`). Would be greateful for a hint.

````cmd
>>> quarto preview "c:/Users/enry/Documents/quatro/hello.qmd" --no-browser --no-watch-inputs

Executing 'hello.ipynb'
  Cell 1/1...ERROR:

An error occurred while executing the following cell:
------------------

import numpy as np

import matplotlib.pyplot as plt

r = np.arange(0, 2, 0.01)
theta = 2 * np.pi * r
fig, ax = plt.subplots(
  subplot_kw = {'projection': 'polar'}
)
ax.plot(theta, r)
ax.set_rticks([0.5, 1, 1.5, 2])
ax.grid(True)
plt.show()
------------------

---------------------------------------------------------------------------      
ModuleNotFoundError                       Traceback (most recent call last)      
<ipython-input-2-9ce6f0b50c5e> in <module>
      1 import numpy as np
      2 
----> 3 import matplotlib.pyplot as plt
      4 
      5 r = np.arange(0, 2, 0.01)

~\anaconda3\lib\site-packages\matplotlib\pyplot.py in <module>
   2334     dict.__setitem__(rcParams, "backend", rcsetup._auto_backend_sentinel)
   2335 # Set up the backend.
-> 2336 switch_backend(rcParams["backend"])
   2337 
   2338 # Just to be safe.  Interactive mode can be turned on without

~\anaconda3\lib\site-packages\matplotlib\pyplot.py in switch_backend(newbackend)
    274     backend_name = cbook._backend_module_name(newbackend)
    275
--> 276     class backend_mod(matplotlib.backend_bases._Backend):
    277         locals().update(vars(importlib.import_module(backend_name)))
    278

~\anaconda3\lib\site-packages\matplotlib\pyplot.py in backend_mod()
    275
    276     class backend_mod(matplotlib.backend_bases._Backend):
--> 277         locals().update(vars(importlib.import_module(backend_name)))
    278
    279     required_framework = _get_required_interactive_framework(backend_mod)

~\anaconda3\lib\importlib\__init__.py in import_module(name, package)
    125                 break
    126             level += 1
--> 127     return _bootstrap._gcd_import(name[level:], package, level)
    128
    129
```

**Comment by cscheid** [comment]
What's happening here:

```python
import matplotlib.pyplot as plt
```

is an import error:

```
---------------------------------------------------------------------------      
ModuleNotFoundError                       Traceback (most recent call last)      
<ipython-input-2-9ce6f0b50c5e> in <module>
      1 import numpy as np
      2 
----> 3 import matplotlib.pyplot as plt
      4 
      5 r = np.arange(0, 2, 0.01)
```

That probably means that the version of python that's visible to the command-line version of `quarto preview` is not the same version that VS code is seeing. You will need to check your Python installations and configuration to make sure, for example, that the PATH variables are pointing to the right places, etc.

**Reply by enryH** [reply]
Thanks! I also found a hint that there is an issue with python kernels in Anaconda environments, see [here](https://quarto.org/docs/computations/python.html#kernel-selection): Does this mean I need to install a new tool? Other places ([issue 147](https://github.com/quarto-dev/quarto-web/issues/147)) it says that recent version should pick up the kernel from the VS Code python extension...

It works all fine with papermill and nbconvert. Can I install quatro separately (so not globally) in an environment? I only need it for one or two, not everywhere.

**Reply by cscheid** [reply]
> Can I install quarto separately (so not globally) in an environment?

To clarify: quarto is not a python tool, so you'll never install it in a virtual environment. With that said, you can definitely have a local install of quarto (at least on Unix! I'm not a Windows user). @cderv What do you think might be happening here?

**Reply by enryH** [reply]
I also have other non-python tools in my anaconda virtual environments. It is rather a tool for having a separate installation of binaries and packages.

**Comment by enryH** [comment]
And it says that starting the kernel works: `Starting python3 kernel...Done`  (which is the one I select also for execution in VS Code)

**Reply by enryH** [reply]
So it finds `matplotlib`, but not the backend it requires: Does quarto need a special backend for matplotlib or is it agnostic to it?

**Reply by enryH** [reply]
I found a more extensive log file of quarto ( `QUARTO_LOG`) and there it specifies the backend: `ModuleNotFoundError: No module named 'matplotlib_inline'`

**Reply by enryH** [reply]
The python exectuable is set to my base conda environment. So this is always used and then a `kernel` is started?

I was recommended to have a separate jupyter installation in each of my conda environments. So I don't use the kernels from my base environment.

**Reply by cscheid** [reply]
> Starting python3 kernel...Done

> So it finds matplotlib, but not the backend it requires: Does quarto need a special backend for matplotlib or is it agnostic to it?

To clarify: the `Starting python3 kernel` message does not mean that `matplotlib` was found. It only means that _a_ jupyter python3 kernel has started.

> The python exectuable is set to my base conda environment. So this is always used and then a kernel is started?

quarto uses the Python environment available from the PATH it sees in the shell it's executing. If VS Code changes the PATH environment variable (which it often does), then that might the source of your mismatch.

**Reply by enryH** [reply]
I'll check `PATH`s again the next time I try, but for now I fall back to `nbconvert`. Thanks for trying to help though.

**Reply by enryH** [reply]
(and I tried it also using jupter lab instruction without success)
